v2/pkg/engine/datasource/grpc_datasource/grpc_datasource_test.go

```diff
@@ -0,0 +1,1054 @@
+	require.Equal(t, "Test Product", resp.Data.ComplexFilterType[0].Name)
+}
+
+func Test_DataSource_Load_WithMockService_WithResponseMapping(t *testing.T) {
+	// 1. Start a real gRPC server with our mock implementation
+	lis := bufconn.Listen(1024 * 1024)
+
+	// Create a new gRPC server
+	server := grpc.NewServer()
+
+	// Register our mock service implementation
+	mockService := &grpctest.MockService{}
+	productv1.RegisterProductServiceServer(server, mockService)
+
+	// Start the server in a goroutine
+	go func() {
+		if err := server.Serve(lis); err != nil {
+			t.Errorf("failed to serve: %v", err)
+		}
+	}()
+
+	// Clean up the server when the test completes
+	defer server.Stop()
+
+	// Create a buffer-based dialer
+	bufDialer := func(context.Context, string) (net.Conn, error) {
+		return lis.Dial()
+	}
+
+	// Connect using bufconn dialer
+	// see https://github.com/grpc/grpc-go/issues/7091
+	// nolint: staticcheck
+	conn, err := grpc.Dial(
+		"bufnet",
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithContextDialer(bufDialer),
+		grpc.WithLocalDNSResolution(),
+	)
+	require.NoError(t, err)
+	defer conn.Close()
+
+	// 3. Set up GraphQL query and schema
+	query := `query ComplexFilterTypeQuery($filter: ComplexFilterTypeInput!) { complexFilterType(filter: $filter) { id name } }`
+	variables := `{"variables":{"filter":{"filter":{"name":"HARDCODED_NAME_TEST","filterField1":"value1","filterField2":"value2"}}}}`
+
+	// Parse the GraphQL schema
+	schemaDoc := grpctest.MustGraphQLSchema(t)
+
+	// Parse the GraphQL query
+	queryDoc, report := astparser.ParseGraphqlDocumentString(query)
+	if report.HasErrors() {
+		t.Fatalf("failed to parse query: %s", report.Error())
+	}
+	compiler, err := NewProtoCompiler(grpctest.MustProtoSchema(t), nil)
+	if err != nil {
+		t.Fatalf("failed to compile proto: %v", err)
+	}
+
+	// 4. Create a datasource with the real gRPC client connection
+	ds, err := NewDataSource(conn, DataSourceConfig{
+		Operation:    &queryDoc,
+		Definition:   &schemaDoc,
+		SubgraphName: "Products",
+		Compiler:     compiler,
+		Mapping: &GRPCMapping{
+			Fields: map[string]FieldMap{
+				"Query": {
+					"complexFilterType": {
+						TargetName: "complex_filter_type",
+					},
+				},
+				"FilterType": {
+					"name": {
+						TargetName: "name",
+					},
+					"filterField1": {
+						TargetName: "filter_field_1",
+					},
+					"filterField2": {
+						TargetName: "filter_field_2",
+					},
+				},
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	// 5. Execute the query through our datasource
+	output := new(bytes.Buffer)
+
+	// Format the input with query and variables
+	inputJSON := fmt.Sprintf(`{"query":%q,"body":%s}`, query, variables)
+	t.Logf("Input JSON: %s", inputJSON)
+
+	err = ds.Load(context.Background(), []byte(inputJSON), output)
+	require.NoError(t, err)
+
+	// Print the response for debugging
+	responseData := output.String()
+	t.Logf("Response: %s", responseData)
```

Comment:
do we still need to output debug logs?

Reply:
No that was a leftover but I'm asserting the response.